addons/sale_crm/models/crm_team.py

```diff
@@ -42,3 +42,51 @@ def _extra_sql_conditions(self):
         if self.use_opportunities and self._context.get('in_sales_app'):
             return "AND state in ('sale', 'done', 'pos_done')"
         return super(CrmTeam,self)._extra_sql_conditions()
+
+    def _merge_leads(self, leads, values):
+        """ Remove opportunities or leads from ["leads_dups_dict"] that
+        have sale orders and the company of the sale order is not the same
+        as the company of the head lead, its sale order, or current company.
+
+        :param leads: recordset of leads to assign to current team;
+        :param values: dictionary in the following form
+                       {
+                            'leads_assigned': crm.lead(),
+                            'leads_dups_dict': {crm.lead(): crm.lead()}
+                       };
+        :return: Opportunities or leads that have sale orders, and the company
+        of the sale order is not the same as the company of the head lead, its
+        sale order, or current company.
+        (
+            {'leads_with_different_company': opportunities}
+        )
+        """
+        for lead in leads.filtered(lambda lead: lead in values['leads_dups_dict']):
+            leads_dups_dict = values['leads_dups_dict'][lead]
+            if leads_dups_dict.order_ids \
+               and (len(leads_dups_dict.order_ids.company_id) > 1 or len(leads_dups_dict.company_id) > 1):
+
+                opportunities = leads_dups_dict._sort_by_confidence_level(reverse=True)
+                head_opportunity = opportunities[0]
+                current_company = head_opportunity.company_id or head_opportunity.order_ids.company_id or self.env.company
+                leads_with_so_different_company = opportunities.filtered(
+                    lambda lead: lead.order_ids and lead.order_ids.company_id != current_company
+                )
+                if len(leads_dups_dict - leads_with_so_different_company) >= 2:
+                    if leads in leads_with_so_different_company:
+                        leads = head_opportunity
```

Comment:
This will not work with multiple leads. I would have a new_leads record_set before the loop and use that to call ``super()``
```suggestion
                    if lead in leads_with_so_different_company:
                        new_leads.append(head_opportunity)
```